src/runtime_src/core/edge/drm/zocl/zocl_kds.c

```diff
@@ -24,6 +25,144 @@ module_param(kds_echo, int, (S_IRUGO|S_IWUSR));
 MODULE_PARM_DESC(kds_echo,
 		 "enable KDS echo (0 = disable (default), 1 = enable)");
 
+static inline void
+zocl_ctx_to_info(struct drm_zocl_ctx *args, struct kds_ctx_info *info)
+{
+	if (args->cu_index == ZOCL_CTX_VIRT_CU_INDEX)
+		info->cu_idx = CU_CTX_VIRT_CU;
+	else
+		info->cu_idx = args->cu_index;
+
+	/* Ignore ZOCL_CTX_SHARED bit if ZOCL_CTX_EXCLUSIVE bit is set */
+	if (args->flags & ZOCL_CTX_EXCLUSIVE)
+		info->flags = CU_CTX_EXCLUSIVE;
+	else
+		info->flags = CU_CTX_SHARED;
+}
+
+static int
+zocl_add_context(struct drm_zocl_dev *zdev, struct kds_client *client,
+		 struct drm_zocl_ctx *args)
+{
+	struct kds_ctx_info info;
+	void *uuid_ptr = (void *)(uintptr_t)args->uuid_ptr;
+	uuid_t *id;
+	int ret;
+
+	id = vmalloc(sizeof(uuid_t));
+	if (!id)
+		return -ENOMEM;
+
+	ret = copy_from_user(id, uuid_ptr, sizeof(uuid_t));
+	if (ret)
+		goto out;
+
+	mutex_lock(&client->lock);
+	if (!client->num_ctx) {
+		ret = zocl_lock_bitstream(zdev, id);
+		if (ret)
+			goto out;
+		client->xclbin_id = vzalloc(sizeof(*id));
+		if (!client->xclbin_id) {
+			ret = -ENOMEM;
+			goto out;
+		}
+		uuid_copy(client->xclbin_id, id);
+	}
+
+	/* Bitstream is locked. No one could load a new one
+	 * until this client close all of the contexts.
+	 */
+	zocl_ctx_to_info(args, &info);
+	ret = kds_add_context(client, &info);
+
+out:
+	if (!client->num_ctx) {
+		vfree(client->xclbin_id);
+		client->xclbin_id = NULL;
+		(void) zocl_unlock_bitstream(zdev, id);
+	}
+	vfree(id);
+	mutex_unlock(&client->lock);
```

Comment:
Nit, you may want to put mutex_unlock before vfree() and make a sequence pair 
alloc
  mutex_lock
  mutex_unlock
free

Reply:
fixed